### PR TITLE
core: make messaging optional

### PR DESCRIPTION
**Pull request by kaspar030**
In order to enhance RIOT's modularity, make messaging optional.

When not used but compiled in, messaging uses a whopping 24b of code on 32bit platforms, but doubles the size of each tcb (16b -> 40b).

Not sure the savings are worth the effort, but who knows...

(Added msg to DEFAULT_MODULES, so this will not affect anything but those brave souls who compile RIOT with ```DISABLE_MODULE += msg```).

### CODE REVIEW

**Comment by OlegHahm** [comment]
Interesting

**Comment by IldarValiev** [comment]
@kaspar030, may be it will be better to add new flag DISABLE_MESSAGING instead of MODULE_MSG - it makes more sense.

I think is important PR, because not in every case we need messaging, especially if RIOT controls simple sensor node.

**Comment by OlegHahm** [comment]
Sorry, this got lost somehow. Could you rebase, @kaspar030?

**Comment by kaspar030** [comment]
- rebased

File: tests/sizeof_tcb/Makefile [hunk 1]
```diff
@@ -1,4 +1,7 @@
 APPLICATION = sizeof_tcb
 include ../Makefile.tests_common
 
+# optionally disable messaging
+#DISABLE_MODULE += msg
```
Comment (OlegHahm) [inline tests/sizeof_tcb/Makefile]: Wouldn't it make sense to comment this in here?

Maybe you could add this snippet also to `dist/Makefile`?

File: Makefile.defaultmodules [hunk 2]
```diff
@@ -1,3 +1,3 @@
-DEFAULT_MODULE += board cpu core sys
+DEFAULT_MODULE += board cpu core sys msg
```
Comment (OlegHahm) [inline Makefile.defaultmodules]: I would swap the order (makes more sense semantically and alphabetical).

**Comment by OlegHahm** [comment]
I'm not sure, I see a concrete use case for that right now, but increasing modularity and making more features optional seems always a good idea if the overhead is manageable - which is the case for this change.

File: tests/sizeof_tcb/Makefile (hunk 1 above)
Comment (kaspar030) [inline tests/sizeof_tcb/Makefile]: > Wouldn't it make sense to comment this in here?

hm?

Comment (OlegHahm) [inline tests/sizeof_tcb/Makefile]: I meant: why not indeed disabling the `msg` module here?

Comment (kaspar030) [inline tests/sizeof_tcb/Makefile]: Ah. Nope, IMHO sizeof_tcb should show the tcb size in the general case. disabling messaging seems more on the special side.

Comment (OlegHahm) [inline tests/sizeof_tcb/Makefile]: Ah, yes, sorry, confused with the minimal example.

**Comment by kaspar030** [comment]
I've re-done the PR, it felt pointless to create a whole directory for msg, now I'm using ifdef within msg.c.

I've moved thread_print_msg_queue() to msg.c and renamed it to msg_queue_print(). @authmillenon, do you approve?

**Comment by authmillenon** [comment]
I would prefer creating a new directory. Other then that `msg_queue_print()` is okay. :-)

**Comment by authmillenon** [comment]
(especially since compilers will complain about the empty compile unit).

**Comment by authmillenon** [comment]
((I saw enough of this type of "modulization" in lwip and emb6 that I know I don't want to have that in RIOT ;-)))

**Comment by kaspar030** [comment]
> I would prefer creating a new directory.

why?

> (especially since compilers will complain about the empty compile unit).
AFAIK, not if there's a single valid source line (e.g., headers).

**Comment by authmillenon** [comment]
I gave already a bunch of reasons but here are more:

 * it's ugly
 * it gives a bad example and will enable future contributors to provide modules the same way
 * it's hard to assign binary data to modules (not in this example, but in lwIP - which uses *only* this kind of moduling - it definitely is a problem)

Compared to that:
 * moving it to a directory isn't hard ;-)

**Comment by kaspar030** [comment]
Martine, how is it possible that we end up having completely opposite opinions on so many matters? ;)

> I gave already a bunch of reasons but here are more:
No you didn't. You gave one, which I think is invalid.

>     it's ugly

Subjective. I don't agree. At least it is a lot less ugly than a subdirectory.

>     it gives a bad example and will enable future contributors to provide modules the same way

this is not the first module to do this.

>     it's hard to assign binary data to modules (not in this example, but in lwIP - which uses only this kind of moduling - it definitely is a problem)

Meaning - what?

> Compared to that:
> 
>     moving it to a directory isn't hard ;-)

It was in a subdirectory before the last commit(s). That looked ugly and completely unnecessary.

Mind you - the whole idea of disabling messaging is more an excercise of "because we can".
If it adds a pseudomodule and an #ifdef'd code block, fine. If it adds another directory (and thus becomes the only core .c file not in /core), it feels too intrusive to me for what it can do.

One-file modules in their own directory always seemed like a deficiency of the build system anyways.

**Comment by OlegHahm** [comment]
Hm, difficult thing. On the one hand, having a subdirectory is kinda ugly, particular since it would be the only subfolder in `core`. On the other hand, this ifdef bracket around basically a whole file is kinda ugly, too. But finally, we cannot do this without some ifdefs anyway, so I'm willing to accept the current solution.

**Comment by authmillenon** [comment]
Well if #4919 is imported and also added as a directory `msg` wouldn't be the only directory in core ;-). Also: I like directories they are not ugly :P.

**Comment by authmillenon** [comment]
But fine. If I'm the only one that diggs subdirectories I'm fine with #ifdef blocks

**Comment by authmillenon** [comment]
> this is not the first module to do this.

Which other pseudo-module does this? I see only dependency-collecting modules, modules that are only header files, common API modules, and "pseudo-submodules" (modules that add additional functionality to an existing module) in Makefile.pseudomodules. There is no module that is exactly like a normal module, but is a pseudo-module because the complete code for is in an `#ifdef`.

**Comment by OlegHahm** [comment]
Needs a rebase.

**Comment by haukepetersen** [comment]
I like the PR and I agree with @kaspar030, that I don't really like putting it into a separate folder. The only thing I think could make sense, is to rename the module to `core_msg`, to make clear that this module is part of core. @authmillenon can you live with this?

so 1st ACK

**Comment by authmillenon** [comment]
Yeah.

**Comment by kaspar030** [comment]
- rebased, now waiting for #4103 as some lines are changed by both

**Comment by kaspar030** [comment]
Would someone give a second ACK?

**Comment by kaspar030** [comment]
green! :)

File: dist/Makefile [hunk 3]
```diff
@@ -36,6 +36,11 @@ QUIET ?= 1
 #USEMODULE += posix
 #USEMODULE += xtimer
 
+# If your application is very simple and doesn't use modules that use
+# messaging, it can be disabled to save some memory:
+
+#DISABLE_MODULE += msg
```
Comment (authmillenon) [inline dist/Makefile]: core_msg

**Comment by authmillenon** [comment]
One minor doc thingy. Otherwise, fine... ACK

**Comment by kaspar030** [comment]
Thx! addressed & amended.

**Pull request merged**